cmd/tailscale/cli/file.go

```diff
@@ -144,30 +162,78 @@ func runCp(ctx context.Context, args []string) error {
 				return errors.New("directories not supported")
 			}
 			contentLength = fi.Size()
-			fileContents = io.LimitReader(f, contentLength)
+			fileContents = &countingReader{io.LimitReader(f, contentLength), 0}
 			if name == "" {
 				name = filepath.Base(fileArg)
 			}
 
 			if envknob.Bool("TS_DEBUG_SLOW_PUSH") {
-				fileContents = &slowReader{r: fileContents}
+				fileContents = &countingReader{&slowReader{r: fileContents}, 0}
 			}
 		}
 
 		if cpArgs.verbose {
 			log.Printf("sending %q to %v/%v/%v ...", name, target, ip, stableID)
 		}
+
+		var (
+			done = make(chan struct{}, 1)
+			wg   sync.WaitGroup
+		)
+		if isatty.IsTerminal(os.Stderr.Fd()) {
+			go func() {
+				defer wg.Done()
+				var lastBytesRead uint64
+				for {
+					select {
+					case <-done:
+						fmt.Fprintln(os.Stderr)
+						return
+					case <-time.After(time.Second):
+						n := fileContents.ReadBytes()
+						contentLengthStr := "???"
+						if contentLength > 0 {
+							contentLengthStr = fmt.Sprint(contentLength / 1024)
+						}
+
+						fmt.Fprintf(os.Stderr, "\r\x1b[K%s\t\t%s", padTruncateString(name, 64), padTruncateString(fmt.Sprintf("%d/%s kb", n/1024, contentLengthStr), 16))
+						if contentLength > 0 {
+							fmt.Fprintf(os.Stderr, "\t%.02f%%", float64(n)/float64(contentLength)*100)
+						} else {
+							fmt.Fprintf(os.Stderr, "\t-------%%")
+						}
+						if lastBytesRead > 0 {
+							fmt.Fprintf(os.Stderr, "\t%d kb/s", (n-lastBytesRead)/1024)
+						} else {
+							fmt.Fprintf(os.Stderr, "\t-------")
+						}
+						lastBytesRead = n
+					}
+				}
+			}()
+			wg.Add(1)
+		}
+
 		err := localClient.PushFile(ctx, stableID, contentLength, name, fileContents)
 		if err != nil {
 			return err
 		}
 		if cpArgs.verbose {
 			log.Printf("sent %q", name)
 		}
+		done <- struct{}{}
+		wg.Wait()
 	}
 	return nil
 }
 
+func padTruncateString(str string, truncateAt int) string {
+	if len(str) <= truncateAt {
+		return str + strings.Repeat(" ", truncateAt-len(str))
+	}
+	return str[:truncateAt-1] + "…"
```

Comment:
this truncates in the middle of a utf-8 sequence

Reply:
Made it respect code-point boundaries.